gh-141004: Document symbol visibility macros (PyAPI_DATA, Py_EXPORTED_SYMBOL, Py_LOCAL_SYMBOL,Py_IMPORTED_SYMBOL) (#143508)
python/cpython: Doc/c-api/intro.rst

   Equivalent to :c:macro:`Py_LOCAL` but additionally requests the function
   be inlined.

.. c:macro:: Py_LOCAL_SYMBOL

   Macro used to declare a symbol as local to the shared library (hidden).
   It ensures the symbol is not exported.
   On platforms with visibility support, it

>    expands to ``__attribute__((visibility("hidden")))``.

Reviewer: This is platform-specific, and not really guaranteed everywhere.
```suggestion
   On supported platforms, it ensures the symbol is not exported.

   On compatible versions of GCC/Clang, it
   expands to ``__attribute__((visibility("hidden")))``.
```

      sizeof(array) / sizeof((array)[0])

.. c:macro:: Py_EXPORTED_SYMBOL

   Macro used to declare a symbol (function or data) as exported from a shared library.
   On Windows, this expands to ``__declspec(dllexport)``.
   On other platforms with visibility support, it
   expands to ``__attribute__((visibility("default")))``.

.. c:macro:: Py_IMPORTED_SYMBOL

   Macro used to declare a symbol as imported from a shared library.
   On Windows, this expands to ``__declspec(dllimport)``.

>    On other platforms, it is usually empty or standard visibility.

Reviewer: “from a shared library” is not true: these depend on whether *CPython*'s own C API is provided as a shared library.

The docs should say that these are for defining the C API itself; users shouldn't use them for their own symbols.

> .. c:macro:: PyAPI_DATA(type)

Reviewer: If we document this, we should also document `PyAPI_FUNC`.
Again, this is for defining the C API itself; users shouldn't use it for their own data.

Reviewer: This should be similar to the docs for `PyAPI_FUNC`.

   Macro used to declare a public global variable.
   It expands to ``extern Py_EXPORTED_SYMBOL type`` or ``extern Py_IMPORTED_SYMBOL type``
   depending on whether the core is being built or used.

   Example usage::

      PyAPI_DATA(PyObject *) _Py_NoneStruct;

> .. c:macro:: Py_LOCAL_SYMBOL

Reviewer: Could you move this next to the other `Py_LOCAL` macros, and ideally match their style?

Reviewer: I guess `Py_ALWAYS_INLINE`, `Py_DEPRECATED`, `Py_LOCAL`, `Py_LOCAL_INLINE`, all of these, and possibly `Py_UNUSED`, could be moved to a new section. But that can be in a new PR.

      sizeof(array) / sizeof((array)[0])

.. c:macro:: Py_EXPORTED_SYMBOL

   Macro used to declare a symbol (function or data) as exported.
   On Windows, this expands to ``__declspec(dllexport)``.

>    On other platforms with visibility support, it

Reviewer: Not sure what you mean “platforms with visibility support”. Would it work to mention the compiler directly?
```suggestion
   On compatible versions of GCC/Clang, it
```

(same below)

   expands to ``__attribute__((visibility("default")))``.
   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: Py_IMPORTED_SYMBOL

   Macro used to declare a symbol as imported.
   On Windows, this expands to ``__declspec(dllimport)``.

>    On other platforms, it is usually empty or standard visibility.

Reviewer: I think we can remove this.
```suggestion
```

As I see it, the main reason for the expansions is so that the reader can look up the behaviour in compiler docs.

   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: PyAPI_FUNC(type)

   Macro used by CPython to declare a function as part of the C API.
   Its expansion depends on the platform and build configuration.
   This macro is intended for defining CPython's C API itself;

>    extension modulesshould not use it for their own symbols.

Reviewer: ```suggestion
   extension modules should not use it for their own symbols.
```

.. c:macro:: PyAPI_DATA(type)

   Macro used to declare a public global variable.
   It expands to ``extern Py_EXPORTED_SYMBOL type`` or ``extern Py_IMPORTED_SYMBOL type``
   depending on whether the core is being built or used.

   Example usage::

>       PyAPI_DATA(PyObject *) _Py_NoneStruct;

Reviewer 2: I would prefer to show the example of a public variable:

```suggestion
      PyAPI_DATA(const unsigned long) Py_Version;
```

      sizeof(array) / sizeof((array)[0])

.. c:macro:: Py_EXPORTED_SYMBOL

   Macro used to declare a symbol (function or data) as exported.
   On Windows, this expands to ``__declspec(dllexport)``.
   On compatible versions of GCC/Clang, it
   expands to ``__attribute__((visibility("default")))``.
   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: Py_IMPORTED_SYMBOL

   Macro used to declare a symbol as imported.
   On Windows, this expands to ``__declspec(dllimport)``.
   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: PyAPI_FUNC(type)

   Macro used by CPython to declare a function as part of the C API.
   Its expansion depends on the platform and build configuration.
   This macro is intended for defining CPython's C API itself;
   extension modules should not use it for their own symbols.

.. c:macro:: PyAPI_DATA(type)

   Macro used to declare a public global variable.
   It expands to ``extern Py_EXPORTED_SYMBOL type`` or ``extern Py_IMPORTED_SYMBOL type``

>    depending on whether the core is being built or used.

Reviewer 2: I would also add:

"This macro is intended for defining CPython's C API itself; extension modules should not use it for their own symbols."

Author: gonna push a PR for the manual changes in a bit yes, thank you for the corrections

      sizeof(array) / sizeof((array)[0])

.. c:macro:: Py_EXPORTED_SYMBOL

   Macro used to declare a symbol (function or data) as exported.
   On Windows, this expands to ``__declspec(dllexport)``.
   On compatible versions of GCC/Clang, it
   expands to ``__attribute__((visibility("default")))``.
   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: Py_IMPORTED_SYMBOL

   Macro used to declare a symbol as imported.
   On Windows, this expands to ``__declspec(dllimport)``.
   This macro is for defining the C API itself; extension modules should not use it.

.. c:macro:: PyAPI_FUNC(type)

   Macro used by CPython to declare a function as part of the C API.
   Its expansion depends on the platform and build configuration.
   This macro is intended for defining CPython's C API itself;
   extension modules should not use it for their own symbols.

.. c:macro:: PyAPI_DATA(type)

   Macro used to declare a public global variable.
   It expands to ``extern Py_EXPORTED_SYMBOL type`` or ``extern Py_IMPORTED_SYMBOL type``

>    depending on whether the core is being built or used.

Reviewer: Could you use the same wording as for `PyAPI_FUNC`?
```suggestion
   Macro used by CPython to declare a public global variable as part of the C API.
   Its expansion depends on the platform and build configuration.
```

   This macro is intended for defining CPython's C API itself;
   extension modules should not use it for their own symbols.

   Example usage::

>       PyAPI_DATA(const unsigned long) Py_Version;

Reviewer: We're saying you shouldn't use this, so let's not give examples.
```suggestion
```
